### ISSUE TITLE
Track : Nullable string fields

### ISSUE THREAD

**Issue by sandreas** [Zeugma440/atldotnet#169]
## The problem

In https://github.com/sandreas/tone/issues/15#issue-1292428672 I try to remove metadata properties completely which does not work as expected.

Unfortunately I did not find a way to completely remove a property - just set it to empty string, which is not the same in my opinion.

My question: Is there a way to get rid of an existing string property - so REALLY removing it?

## Code To Reproduce Issue [ Good To Have ]
```c#
// ATL.Settings.NullAbsentValues = true; // does not change the behaviour in any way (which is good)
var track = new Track("a-file.m4b");
track.Album = null; // does nothing instead of removing the property, which I had expected
// track.Album = ""; // sets the property to empty string (and does not remove it)
track.Save();
```

I would:
- remove the property on setting it to `null` (regardless of the properties type, e.g. `Track.RecordingDate = null` should remove it)
- set the property to empty string on strings with `Track.Album = "";` (as it is now)
- introduce a new setting `Settings.TreatDefaultValuesAsNull = true;`
  - keep the current behaviour on `Settings.TreatDefaultValuesAsNull = true;`, e.g. `Track.TrackNumber = 0;` removes it?
  - set the property to the given value on `Settings.TreatDefaultValuesAsNull = false;`, e.g. `Track.TrackNumber = 0;`

What do you think?

**Comment by Zeugma440** [comment]
With the current API : 
- To completely remove a non-`string` field, you can :
  - Either set `Settings.NullAbsentValues = true` and set the field to `null`
  - Or keep the default of `Settings.NullAbsentValues` and set the field to its default value (`0` for numbers, `DateTime.MinValue` for dates)
- To completely remove a `string` field, you should set the field to `""` (empty string)
- As a consequence to this convention, it is not possible to set a field to the value `""`

I have not forgotten you've made a convincing case to evolve the API to allow nullable strings. 
I just need some time for this to happen, as it is a major evolution. For now, please cope with setting empty strings. 

I'll keep that issue open as a reminder. Thanks for your understanding~

**Comment by sandreas** [comment]
> With the current API :

Thank you for clarifying this :-) I think this description would also make a decent FAQ entry, that can be updated as soon as this has been changed.

**Comment by sandreas** [comment]
So to summarize, currently, we have the following types with the according values to remove the value:

- `int` (e.g. `DiscNumber`) => `0`
- `string` (e.g. `Album`) => `""`
- `DateTime` (e.g. `PublishingDate`) => `DateTime.MinValue`
- `float` (e.g. `Popularity`) => `0.0f`

But what about these? Are my assumptions correct? (especially about `Lyrics`):
- `LyricsInfo` (`Lyrics`) => `null`
- `IList<ChapterInfo>` ( `Chapters`) => `new List<ChapterInfo>()` or `Track.Chapters.Clear()`
- `IList<PictureInfo>` (`EmbeddedPictures`) => `new List<PictureInfo>()` or `Track.EmbeddedPictures.Clear()`
- `IDictionary<string, string>` (`AdditionalFields`) => `new Dictionary<string, string>();` or `Track.AdditionalFields.Clear()`

**Comment by Zeugma440** [comment]
Let me rephrase that

**Simple fields**

Field type | Value when `Settings.NullAbsentValues` is `false` | Value when `Settings.NullAbsentValues` is `true`
---|---|---
`int` | `0` | `null`
`float` | `0.0f` | `null`
`DateTime` | `DateTime.MinValue` | `null`
`string` | `""` | `""`

**Complex fields**

Field | Value
---|---
`Lyrics` | Touché... setting to `null` doesn't work ! There's no way to remove lyrics at the moment 😅 
`Chapters` | `new List<ChapterInfo>()` or `Track.Chapters.Clear()`
`EmbeddedPictures` | `new List<PictureInfo>()` or `Track.EmbeddedPictures.Clear()`
`AdditionalFields` | `new Dictionary<string, string>()` or `Track.AdditionalFields.Clear()`

=> During the upcoming API evolution, I'll make it so setting `null` to `Lyrics` actually removes them.

**Comment by sandreas** [comment]
> Touché... setting to null doesn't work ! There's no way to remove lyrics at the moment 😅

Ahaa! I knew I had super nitpicking powers! ;)

> => During the upcoming API evolution, I'll make it so setting null to Lyrics actually removes them.

Great, thank you. I'm really exited about this evolution. If you need my nitpicking to you help with that, just ask =D I'm really impressed with the improvements you did since I started asking the sh.. out of you. You really DO care about your stuff!

**Comment by Zeugma440** [comment]
I haven't had the time to initiate any API evolution, but I did make a small change to be able to remove lyrics.

Setting `Track.Lyrics` to `null` now works as expected since today's v4.23

**Comment by sandreas** [comment]
Cool, thank you. I did not have the time to do anything in the last weeks... although I have already put some effort in an `AdvancedTrack`, its not gonna finish in the next month. Lots of things are happening atm ;)
I'll keep you updated.
